finality-provider/service/fp_instance.go

```diff
@@ -194,52 +194,81 @@ func (fp *FinalityProviderInstance) IsRunning() bool {
-				"the finality-provider received a new block, start processing",
+			channelSize := len(fp.poller.blockInfoChan)
+			fp.logger.Debug("the finality-provider received a new block",
+				zap.Int("poller_channel_size", channelSize),
 				zap.String("pk", fp.GetBtcPkHex()),
 				zap.Uint64("height", b.Height),
 				zap.String("block_hash", hex.EncodeToString(b.Hash)),
 			)
 
-			// check whether the block has been processed before
-			if fp.hasProcessed(b.Height) {
-				continue
-			}
-			// check whether the finality provider has voting power
-			hasVp, err := fp.hasVotingPower(b.Height)
-			if err != nil {
-				fp.reportCriticalErr(err)
-				continue
-			}
-			if !hasVp {
-				// the finality provider does not have voting power
-				// and it will never will at this block
-				fp.MustSetLastProcessedHeight(b.Height)
-				fp.metrics.IncrementFpTotalBlocksWithoutVotingPower(fp.GetBtcPkHex())
-				continue
-			}
-			// check whether the randomness has been committed
-			// the retry will end if max retry times is reached
-			// or the target block is finalized
-			isFinalized, err := fp.retryCheckRandomnessUntilBlockFinalized(b)
-			if err != nil {
-				if !errors.Is(err, ErrFinalityProviderShutDown) {
-					fp.reportCriticalErr(err)
+			// Fetch all available blocks
+			// Note: not all the blocks in the range will have votes cast
+			// due to lack of voting power or public randomness, so we may
+			// have gaps during processing
+			pollerBlocks := []*types.BlockInfo{b}
+			for {
+				select {
+				case b := <-fp.poller.GetBlockInfoChan():
+					fp.logger.Debug(
+						"the finality-provider received a new block",
+						zap.String("pk", fp.GetBtcPkHex()),
+						zap.Uint64("height", b.Height),
+						zap.String("block_hash", hex.EncodeToString(b.Hash)),
+					)
+
+					// check whether the block has been processed before
+					if fp.hasProcessed(b.Height) {
+						continue
+					}
+					// check whether the finality provider has voting power
+					hasVp, err := fp.hasVotingPower(b.Height)
+					if err != nil {
+						fp.reportCriticalErr(err)
+						continue
+					}
+					if !hasVp {
+						// the finality provider does not have voting power
+						// and it will never will at this block
+						fp.MustSetLastProcessedHeight(b.Height)
+						fp.metrics.IncrementFpTotalBlocksWithoutVotingPower(fp.GetBtcPkHex())
+						continue
+					}
+					// check whether the randomness has been committed
+					// the retry will end if max retry times is reached
+					// or the target block is finalized
+					isFinalized, err := fp.retryCheckRandomnessUntilBlockFinalized(b)
+					if err != nil {
+						if !errors.Is(err, ErrFinalityProviderShutDown) {
+							fp.reportCriticalErr(err)
+						}
+						break
+					}
+					// the block is finalized, no need to submit finality signature
+					if isFinalized {
+						fp.MustSetLastProcessedHeight(b.Height)
+						continue
+					}
+
+					pollerBlocks = append(pollerBlocks, b)
+				default:
+					goto processBlocks
```

Comment:
Is there a way to avoid the `goto` statement, as it seems not a good practice.

Here `default` is hit when the channel becomes empty. Can we replace this `goto` with `break` then?

Reply:
replace `goto` statement and use a boolean `fetchMoreBlocks`.